concrete-csprng/src/generators/mod.rs

```
@@ -70,18 +87,40 @@ pub trait RandomGenerator: Iterator<Item = u8> {
     /// Note:
     /// -----
     ///
-    /// A fresh generator can generate 2¹³² bytes. Unfortunately, no rust integer type in is able
+    /// A fresh generator can generate 2¹³² bytes. Unfortunately, no rust integer type is able
     /// to encode such a large number. Consequently [`ByteCount`] uses the largest integer type
     /// available to encode this value: the `u128` type. For this reason, this method does not
     /// effectively return the number of remaining bytes, but instead
     /// `min(2¹²⁸-1, remaining_bytes)`.
     fn remaining_bytes(&self) -> ByteCount;
 
+    /// Returns the number of bytes that have been generated so far. On a fresh generator this
+    /// returns 0.
+    ///
+    /// Note:
+    /// -----
+    ///
+    /// A fresh generator can generate 2¹³² bytes. Unfortunately, no rust integer type is able
+    /// to encode such a large number. Consequently [`ByteCount`] uses the largest integer type
+    /// available to encode this value: the `u128` type. For this reason, this method does not
+    /// effectively return the number of generated bytes, but instead
+    /// `min(2¹²⁸-1, generated_bytes)`.
+    fn generated_bytes_so_far(&self) -> ByteCount;
+
     /// Returns the next byte of the stream, if the generator did not yet reach its bound.
     fn next_byte(&mut self) -> Option<u8> {
         self.next()
     }
 
+    /// Skips n bytes without consuming the generator like `Iterator::skip` does.
+    fn shift_n(&mut self, n: usize) -> Result<(), GeneratorError> {
+        if n > 0 {
+            // nth(n) skips n + 1
+            self.nth(n - 1).ok_or(GeneratorError::ShiftTooLarge(n))?;
+        }
```

Should we return an error when n is not > 0?

---

Hmm that's a good question, I took inspiration from the skip method of the iterator trait and they had something that looked like this

---

Indeed they don't specifically return an error when n is not > 0. I guess it's ok like this then :sweat_smile: